File: 프로그래머스/Level2/Python/기능개발.py
# 코딩테스트 연습 > 스택/큐 > 기능개발 (Level2)
def solution(progresses, speeds):
    answer = []

    while progresses:
        # 진척도를 반영한다.
        for idx, val in enumerate(progresses):
            progresses[idx] = 100 if val + speeds[idx] > 100 else val + speeds[idx]

        # 가장 앞 작업의 진척도가 100이 된 경우,
        if progresses[0] == 100:
            count = 0
            # 연속적으로 진척도가 100인 작업의 개수를 센다
            for idx, val in enumerate(progresses):
                if val == 100:
                    count += 1
                else:
                    break
            answer.append(count)  # 정답에 추가한다.
            progresses = progresses[count:]  # 작업 완료된 부분을 제거한다.
            speeds = speeds[count:]  # 작업 완료된 부분을 제거한다.

    return answer


print(solution([95, 90, 99, 99, 80, 99], [1, 1, 1, 1, 1, 1]))  # 1, 3, 2

File: 프로그래머스/Level2/Javascript/기능개발.js
function solution(progresses, speeds) {
    var answer = [];
    
    while (progresses.length > 0) {
        // 각 작업에 대한 진척도를 반영한다.
        progresses = progresses.map((progress, index) => {
            const nextProgress = progress + speeds[index];
            return nextProgress > 100 ? 100 : nextProgress;
        })
        
        // 가장 첫번째 작업이 완료된 경우,
        if (progresses[0] === 100) {
            // 종료되지 않은 첫번째 작업의 인덱스를 구한다.
            const workInProgressIndex = progresses.findIndex(progress => progress < 100);
            
            // 종료되지 않은 작업이 없다면,
            if (workInProgressIndex === -1) {
                // 현재 Queue에 존재하는 모든 작업의 개수를 추가하고 종료한다. 
                answer.push(progresses.length);
                break;
            }

            // 종료된 작업들을 작업 진도, 작업 속도 배열에서 모두 제거한다.
            answer.push(workInProgressIndex);
            speeds = speeds.slice(workInProgressIndex);
            progresses = progresses.slice(workInProgressIndex);            
        }
    }
    
    return answer;
}